From: "Gabriel Ivăncescu" <gabrielopcode@gmail.com>
Subject: [PATCH 4/6] win32u: Don't consider cloaked windows as foreground candidates.
Message-Id: <d141a274e2a5d1b7d4260e48e6964890f0a11773.1604678337.git.gabrielopcode@gmail.com>
Date: Fri,  6 Nov 2020 18:03:33 +0200
In-Reply-To: <c2d924287aa6ead5a30239d694d834427a61cf3c.1604678337.git.gabrielopcode@gmail.com>
References: <c2d924287aa6ead5a30239d694d834427a61cf3c.1604678337.git.gabrielopcode@gmail.com>

While manually cloaked windows can be set to foreground, they shouldn't be
considered as activation candidates when trying to find a window to activate
(after hiding a window for example).

Signed-off-by: Gabriel Ivăncescu <gabrielopcode@gmail.com>
---
 dlls/win32u/window.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -31,6 +31,7 @@
 #include "ntuser_private.h"
 #include "wine/server.h"
 #include "wine/debug.h"
+#include "wine/server.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(win);
 
@@ -3837,13 +3838,21 @@ UINT win_set_flags( HWND hwnd, UINT set_mask, UINT clear_mask )
  */
 static BOOL can_activate_window( HWND hwnd )
 {
+    DWORD cloaked = 0;
     LONG style;
 
     if (!hwnd) return FALSE;
     style = get_window_long( hwnd, GWL_STYLE );
     if (!(style & WS_VISIBLE)) return FALSE;
     if ((style & (WS_POPUP|WS_CHILD)) == WS_CHILD) return FALSE;
-    return !(style & WS_DISABLED);
+    if (style & WS_DISABLED) return FALSE;
+    SERVER_START_REQ( get_window_cloaked )
+    {
+        req->handle = wine_server_user_handle( hwnd );
+        if (!wine_server_call( req )) cloaked = reply->cloaked;
+    }
+    SERVER_END_REQ;
+    return !cloaked;
 }
 
 /*******************************************************************

-- 
2.39.0

